### PR TITLE
[RFC] Task Middleware: Round Two

### PR DESCRIPTION
**Pull request by gcoonrod**
This PR is my second attempt at implementing some basic task middleware that behaves like the action middleware.

Still Todo:
- [x] Tests
- [x] Docs

### CODE REVIEW

File: initializers/taskTimer.js [hunk 1]
```diff
@@ -0,0 +1,31 @@
+'use strict';
```
Comment (evantahler) [inline initializers/taskTimer.js]: This is a a great example for the docs, but probably doesn't belong checked in :D

File: initializers/tasks.js [hunk 2]
```diff
@@ -297,11 +331,34 @@ module.exports = {
...
+        api.utils.recursiveDirectoryGlob(p).forEach(function(f){
+          api.tasks.loadFile(f, reload);
+        });
       });
-    });
+    }
+
+    api.tasks.addMiddleware = function(middleware, callback){
+      if(!middleware.name){ return callback(new Error('middleware.name is required')); }
+      if(!middleware.priority){ middleware.priority = api.config.general.defaultMiddlewarePriority; }
```
Comment (evantahler) [inline initializers/tasks.js]: Please add a note to that config file that `api.config.general.defaultMiddlewarePriority` now effects both tasks and actions.

File: initializers/tasks.js [hunk 3]
```diff
@@ -297,11 +331,34 @@ module.exports = {
...
+      middleware.priority = Number(middleware.priority);
+      api.tasks.middleware[middleware.name] = middleware;
+      if(middleware.global === true){
+        api.tasks.globalMiddleware.push(middleware.name);
+        api.tasks.globalMiddleware.sort(function(a, b){
```
Comment (evantahler) [inline initializers/tasks.js]: Now that both actions and tasks are using this sort, it should probably be moved in to utils.

**Comment by evantahler** [comment]
This is excellent, and on the right track!

**Comment by evantahler** [comment]
You can now change `node-resque` to version `v2.1.0` and use `worker.performInline` in specHelper to help with the tests. 

https://github.com/taskrabbit/node-resque/releases/tag/v2.1.0

**Comment by gcoonrod** [comment]
Hmm, not sure why tests won't pass on Travis...

File: initializers/specHelper.js [hunk 4]
```diff
@@ -161,6 +162,27 @@ module.exports = {
         api.tasks.tasks[taskName].run(api, params, next);
       };
 
+      api.specHelper.runFullTask = function(taskName, params, next){
+        var options = {
+          connection: {},
+          queues: api.config.tasks.queues || ['default']
+        };
+        var worker = new NR.worker(options, api.tasks.jobs);
```
Comment (evantahler) [inline initializers/specHelper.js]: I think without a connection, it will try to connect to redis in all cases.  Best to pass in redis object (api.redis.clients.tasks) like we do in `initializers/resque` does

Comment (evantahler) [inline initializers/specHelper.js]: This might be the source of your travis trouble

File: initializers/specHelper.js [hunk 5]
```diff
@@ -161,6 +162,27 @@ module.exports = {
...
+        worker.connect(function(error){
+          if(error){
+            return next(error);
+          }
+
+          worker.performInline(taskName, params, function(error, result){
```
Comment (evantahler) [inline initializers/specHelper.js]: Just a shorthand note: what you have written is the same as `worker.performInline(taskName, params, next)`

File: initializers/tasks.js [hunk 6]
```diff
@@ -44,13 +46,45 @@ module.exports = {
...
+
+              if(self.worker.queueObject){
+                self.queueObject = self.worker.queueObject;
+              }else{
+                self.queueObject = self.worker;
+              }
+            };
+
+            if(api.tasks.middleware[m].preProcessor){ plugin.prototype.before_perform = api.tasks.middleware[m].preProcessor; }
+            if(api.tasks.middleware[m].postProcessor){ plugin.prototype.after_perform = api.tasks.middleware[m].postProcessor; }
```
Comment (evantahler) [inline initializers/tasks.js]: You are only using 2/4 of the possible node-resque hooks.  After this is merged in, please add a github issue to also allow for this middleware to modify `api.tasks.enqueue` with the `before_enqueue` and `after_enqueue` hooks as well

Comment (gcoonrod) [inline initializers/tasks.js]: Those are not trivial changes. Why do you want to expose them?

Comment (evantahler) [inline initializers/tasks.js]: Completeness?  Doing so is certainly not part of this PR

**Comment by evantahler** [comment]
GREEN!

File: actions/randomNumber.js [hunk 7]
```diff
@@ -4,6 +4,9 @@ exports.randomNumber = {
   outputExample: {
     randomNumber: 0.123
   },
+  inputs: {
```
Comment (evantahler) [inline actions/randomNumber.js]: I'm confused about why `test` needs to be hacked into params?  You aren't using an action at all in your work, so why does this mater?  Also, there are better ways of getting this in if you need it than polluting an action's definition.

File: test/tasks/middleware.js [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+
+  after(function(done){
+    api.tasks.globalMiddleware = [];
+    actionhero.stop(function(){
+      done();
+    });
+  });
+
+  it('can modify parameters before a task and modify result after task completion', function(done){
+    api.specHelper.runFullTask('runAction', taskParams, function(error, response){
```
Comment (evantahler) [inline test/tasks/middleware.js]: Oh! You are trying to use the 'run action' task.  I think that's an overly complicated task to use... just define a new simple one in-line as part of your test.  See https://github.com/evantahler/actionhero/blob/master/test/core/tasks.js#L18-L78

**Comment by evantahler** [comment]
I think this is good to go! 
Please change the branch you are merging into to the new `v15` branch!
